# Segmenting and Clustering Neighborhoods in Toronto

## Part 1

For this assignment, you will be required to explore and cluster the neighborhoods in Toronto.

1. Start by creating a new Notebook for this assignment.
2. Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe like the one shown below:

<img src = "https://d3c33hcgiwev3.cloudfront.net/imageAssetProxy.v1/7JXaz3NNEeiMwApe4i-fLg_40e690ae0e927abda2d4bde7d94ed133_Screen-Shot-2018-06-18-at-7.17.57-PM.png?expiry=1569715200000&hmac=lcR8JvkhG5CJniViy-x_Nq4wGOOnaIhDu63SxaZsPgk">

3. To create the above dataframe:
<br>a. The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood.
<br>b. Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
<br>c. More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.
<br>d. If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.
<br>e. Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
<br>f. In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.
4. Submit a link to your Notebook on your Github repository. <b>(10 marks)</b>

In [1]:
# Import necessary libraries
import requests
import lxml.html as lh
import pandas as pd

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

#Create a handle 'page' to handle the contents of the website
page = requests.get(url)

#Store the contents of the website under doc
doc = lh.fromstring(page.content)

#Parse data that are stored between <tr>..</tr> of HTML
tr_elements = doc.xpath('//tr')

In [3]:
#Check the length of the first 10 rows
[len(T) for T in tr_elements[:10]]

[3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

Each row has 3 columns. Let us check the headers.

In [4]:
# Parse the first row as our header
tr_elements = doc.xpath('//tr')

#Create empty list
col = []
i = 0

#For each row, store each first element (header) and an empty list
for t in tr_elements[0]:
    i = i + 1
    name = t.text_content()
    print ('%d:"%s"'%(i, name))
    col.append((name, []))

1:"Postcode"
2:"Borough"
3:"Neighbourhood
"


Let us now create a pandas dataframe and store the data in it.

In [5]:
#Since our first row is the header, data is stored on the second row onwards
for j in range(1, len(tr_elements)):
    #T is our j'th row
    T = tr_elements[j]
    
    #If row is not of size 3, the //tr data is not from our table 
    if len(T) != 3:
        break
    
    #i is the index of our column
    i = 0
    
    #Iterate through each element of the row
    for t in T.iterchildren():
        data = t.text_content() 
        
        #Check if row is empty
        if i > 0:
        #Convert any numerical value to integers
            try:
                data = int(data)
            except:
                pass
            
        #Append the data to the empty list of the i'th column
        col[i][1].append(data)
        
        #Increment i for the next column
        i = i + 1

Let us check the length of each column.

In [6]:
# Check the length of each column. Ideally, they should all be the same
[len(C) for (title, C) in col]

[288, 288, 288]

Since they are same, the data import was successful in terms of column assignment.

In [7]:
Dict = {title: column for (title, column) in col}
df = pd.DataFrame(Dict)

Let us check the data in the dataframe.

In [8]:
# Access the top 5 rows of the data frame 
df.head()

Postcode           Borough     Neighbourhood\n
0      M1A      Not assigned      Not assigned\n
1      M2A      Not assigned      Not assigned\n
2      M3A        North York         Parkwoods\n
3      M4A        North York  Victoria Village\n
4      M5A  Downtown Toronto      Harbourfront\n

Let us clean the '\n' from the data.

In [9]:
df = df.replace('\n', '', regex = True)
df.head()

Postcode           Borough   Neighbourhood\n
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

Now let us drop all the cells that do not have a borough assigned.

In [10]:
df.drop(df.index[df['Borough'] == 'Not assigned'], inplace = True)

# Reset the index and dropping the previous index
df = df.reset_index(drop = True)
df.head(10)

Postcode           Borough   Neighbourhood\n
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M5A  Downtown Toronto       Regent Park
4      M6A        North York  Lawrence Heights
5      M6A        North York    Lawrence Manor
6      M7A      Queen's Park      Not assigned
7      M9A         Etobicoke  Islington Avenue
8      M1B       Scarborough             Rouge
9      M1B       Scarborough           Malvern

Let us now combine the neighborhoods from same postcodes and boroughs.

In [11]:
df.columns = ['PostalCode', 'Borough', 'Neighborhood']   #removing \n from last header cell of the header and renaming the headers
df = df.groupby(['PostalCode', 'Borough'])['Neighborhood'].apply(','.join).reset_index()
df.head(10)

PostalCode      Borough                                   Neighborhood
0        M1B  Scarborough                                  Rouge,Malvern
1        M1C  Scarborough           Highland Creek,Rouge Hill,Port Union
2        M1E  Scarborough                Guildwood,Morningside,West Hill
3        M1G  Scarborough                                         Woburn
4        M1H  Scarborough                                      Cedarbrae
5        M1J  Scarborough                            Scarborough Village
6        M1K  Scarborough      East Birchmount Park,Ionview,Kennedy Park
7        M1L  Scarborough                  Clairlea,Golden Mile,Oakridge
8        M1M  Scarborough  Cliffcrest,Cliffside,Scarborough Village West
9        M1N  Scarborough                     Birch Cliff,Cliffside West

Assigning Borough values to the Neighborhoods where value is "Not assigned".

In [12]:
df.loc[df['Neighborhood'] == 'Not assigned', 'Neighborhood'] = df['Borough']

In [13]:
# Check if the Neighborhood for Queen's Park changed 
df[df['Borough'] == 'Queen\'s Park']

PostalCode       Borough  Neighborhood
85        M7A  Queen's Park  Queen's Park

Thus, we have successfully performed the required actions.

In [14]:
# Check the shape of the data frame
df.shape

(103, 3)

## Part  2

Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.

In an older version of this course, we were leveraging the Google Maps Geocoding API to get the latitude and the longitude coordinates of each neighborhood. However, recently Google started charging for their API: http://geoawesomeness.com/developers-up-in-arms-over-google-maps-api-insane-price-hike/, so we will use the Geocoder Python package instead: https://geocoder.readthedocs.io/index.html.

The problem with this Package is you have to be persistent sometimes in order to get the geographical coordinates of a given postal code. So you can make a call to get the latitude and longitude coordinates of a given postal code and the result would be None, and then make the call again and you would get the coordinates. So, in order to make sure that you get the coordinates for all of our neighborhoods, you can run a while loop for each postal code. Taking postal code M5G as an example, your code would look something like this:

import geocoder # import geocoder

initialize your variable to None
lat_lng_coords = None

loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]

Given that this package can be very unreliable, in case you are not able to get the geographical coordinates of the neighborhoods using the Geocoder package, here is a link to a csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data

Use the Geocoder package or the csv file to create the following dataframe:
<img src = "https://d3c33hcgiwev3.cloudfront.net/imageAssetProxy.v1/HZ3jNHNOEeiMwApe4i-fLg_f44f0f10ccfaf42fcbdba9813364e173_Screen-Shot-2018-06-18-at-7.18.16-PM.png?expiry=1569715200000&hmac=8Fo5gEMvxibQbUqInUZAInKb18rxvyrQYQWZAmYFbbA">

<b>Important Note:</b> There is a limit on how many times you can call geocoder.google function. It is 2500 times per day. This should be way more than enough for you to get acquainted with the package and to use it to get the geographical coordinates of the neighborhoods in the Toronto.

Once you are able to create the above dataframe, submit a link to the new Notebook on your Github repository. <b>(2 marks)</b>

Let us first download and import all the needed libraries.

In [15]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

#!conda install -c conda-forge geopy --yes    #uncomment to install geopy library
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

#!conda install -c conda-forge folium = 0.5.0 --yes    #uncomment to install folium library
import folium # plotting library

print('Folium installed')
print('Libraries imported.')

Folium installed
Libraries imported.


In [ ]:
#!conda install -c conda-forge geocoder --yes    #uncomment to install geocoder library
import geocoder # import geocoder

# initialize your variable to None
lat_lng_coords = None

# loop until you get the coordinates
while(lat_lng_coords is None):
    g = geocoder.google('{}, Toronto, Ontario'.format()
    lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]

The above block of code threw an unexpected error. Hence, let us download the data directly from the CSV file provided.

In [16]:
import numpy as np # library to handle data in a vectorized manner

link = "http://cocl.us/Geospatial_data"
df1 = pd.read_csv(link)

df1.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [17]:
df1.shape

(103, 3)

Both the dataframes have the same shape with 103 rows and 3 columns.

In [18]:
df1.columns = ['PostalCode', 'Latitude', 'Longitude']  #renaming the column headers

cols = df1.columns.tolist()
cols

['PostalCode', 'Latitude', 'Longitude']

Now, let us merge the two dataframes with respect to the 'PostalCode' column.

In [19]:
neighborhoods = pd.merge(df, df1, on = 'PostalCode')
neighborhoods.head()

PostalCode      Borough                          Neighborhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.806686   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

## Part 3

Explore and cluster the neighborhoods in Toronto. You can decide to work with only boroughs that contain the word Toronto and then replicate the same analysis we did to the New York City data. It is up to you.

Just make sure:

1. To add enough Markdown cells to explain what you decided to do and to report any observations you make.
2. To generate maps to visualize your neighborhoods and how they cluster together.

Once you are happy with your analysis, submit a link to the new Notebook on your Github repository. <b>(3 marks)</b>

In [20]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


Let us check the data.

In [21]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 11 boroughs and 103 neighborhoods.


Let us now fetch the coordinates of Toronto using geopy.

In [22]:
address = 'Toronto, Canada'
geolocator = Nominatim(user_agent = "can_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.653963, -79.387207.


Let us now create a visualization for the map of Toronto.

In [23]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location = [latitude, longitude], zoom_start = 10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhoods, borough)
    label = folium.Popup(label, parse_html = True)
    folium.CircleMarker(
        [lat, lng],
        radius = 5,
        popup = label,
        color = 'blue',
        fill = True,
        fill_color = '#3186cc',
        fill_opacity = 0.7,
        parse_html = False).add_to(map_toronto)  
    
map_toronto

Now, we will use the Foursquare API to explore the neighborhoods in Toronto! 

In [1]:
# Define Foursquare Credentials and Version
LIMIT = 100

CLIENT_ID = '*****' # your Foursquare ID
CLIENT_SECRET = '*****' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: *****
CLIENT_SECRET:*****


Let us create a function to repeat the same process to all the neighborhoods in Toronto.

In [25]:
def getNearbyVenues(names, latitudes, longitudes, radius = 500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now, write the code to run the above function on each neighborhood and create a new dataframe called toronto_venues.

In [26]:
toronto_venues = getNearbyVenues(names = neighborhoods['Neighborhood'],
                                   latitudes = neighborhoods['Latitude'],
                                   longitudes = neighborhoods['Longitude'],
                                  )

Rouge,Malvern
Highland Creek,Rouge Hill,Port Union
Guildwood,Morningside,West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park,Ionview,Kennedy Park
Clairlea,Golden Mile,Oakridge
Cliffcrest,Cliffside,Scarborough Village West
Birch Cliff,Cliffside West
Dorset Park,Scarborough Town Centre,Wexford Heights
Maryvale,Wexford
Agincourt
Clarks Corners,Sullivan,Tam O'Shanter
Agincourt North,L'Amoreaux East,Milliken,Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview,Henry Farm,Oriole
Bayview Village
Silver Hills,York Mills
Newtonbrook,Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park,Don Mills South
Bathurst Manor,Downsview North,Wilson Heights
Northwood Park,York University
CFB Toronto,Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens,Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West,Riverdale
The Beaches West,Indi

Let us check out the new data frame and its dimensions!

In [27]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  \
0                         Rouge,Malvern              43.806686   
1  Highland Creek,Rouge Hill,Port Union              43.784535   
2  Highland Creek,Rouge Hill,Port Union              43.784535   
3       Guildwood,Morningside,West Hill              43.763573   
4       Guildwood,Morningside,West Hill              43.763573   

   Neighborhood Longitude                            Venue  Venue Latitude  \
0              -79.194353                          Wendy's       43.807448   
1              -79.160497           Chris Effects Painting       43.784343   
2              -79.160497            Royal Canadian Legion       43.782533   
3              -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   
4              -79.188711                G & G Electronics       43.765309   

   Venue Longitude              Venue Category  
0       -79.199056        Fast Food Restaurant  
1       -79.163742  Construction & Landscaping  
2       -79.163085                         Bar  
3       -79.189914                 Pizza Place  
4       -79.191537           Electronics Store

In [28]:
toronto_venues.shape

(2268, 7)

Let's check how many venues were returned for each neighborhood.

In [29]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                                100   
Agincourt                                                               4   
Agincourt North,L'Amoreaux East,Milliken,Steele...                      2   
Albion Gardens,Beaumond Heights,Humbergate,Jame...                     10   
Alderwood,Long Branch                                                   9   
Bathurst Manor,Downsview North,Wilson Heights                          19   
Bayview Village                                                         4   
Bedford Park,Lawrence Manor East                                       23   
Berczy Park                                                            57   
Birch Cliff,Cliffside West                                              4   
Bloordale Gardens,Eringate,Markland Wood,Old Bu...                      6   
Brockton,Exhibition Place,Parkdale Village                             21   
Business Reply Mail Processing Centre 969 Eastern                      19   
CFB Toronto,Downsview East                                              2   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                     16   
Cabbagetown,St. James Town                                             43   
Caledonia-Fairbanks                                                     5   
Canada Post Gateway Processing Centre                                  11   
Cedarbrae                                                               7   
Central Bay Street                                                     86   
Chinatown,Grange Park,Kensington Market                               100   
Christie                                                               16   
Church and Wellesley                                                   90   
Clairlea,Golden Mile,Oakridge                                          10   
Clarks Corners,Sullivan,Tam O'Shanter                                  12   
Cliffcrest,Cliffside,Scarborough Village West                           3   
Cloverdale,Islington,Martin Grove,Princess Gard...                      1   
Commerce Court,Victoria Hotel                                         100   
Davisville                                                             36   
Davisville North                                                        8   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                     16   
Del Ray,Keelesdale,Mount Dennis,Silverthorn                             5   
Design Exchange,Toronto Dominion Centre                               100   
Don Mills North                                                         5   
Dorset Park,Scarborough Town Centre,Wexford Hei...                      6   
Dovercourt Village,Dufferin                                            15   
Downsview Central                                                       3   
Downsview Northwest                                                     4   
Downsview West                                                          6   
Downsview,North Park,Upwood Park                                        4   
East Birchmount Park,Ionview,Kennedy Park                               5   
East Toronto                                                            3   
Emery,Humberlea                                                         2   
Fairview,Henry Farm,Oriole                                             66   
First Canadian Place,Underground city                                 100   
Flemingdon Park,Don Mills South                                        23   
Forest Hill North,Forest Hill West                                      4   
Glencairn                                                               4   
Guildwood,Morningside,West Hill                                         7   
Harbord,University of Toronto                                          35   
Harbourfront East,Toronto Islands,Union Station 

Let's find out the number of unique categories for venues.

In [30]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 275 uniques categories.


### Analyzing Neighborhoods

In [31]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0            0                  0                  0        0   
1            0                  0                  0        0   
2            0                  0                  0        0   
3            0                  0                  0        0   
4            0                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    1   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0               0                 0                 0                   0   
1               0                 0                 0                   0   
2               0                 0                 0                   0   
3               0                 0                 0                   0   
4               0                 0                 0                   0   

   Beach  Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  \
0      0                0         0           0                   0   
1      0                0         0           0                   0   
2      0                0         0           0                   0   
3      0                0         0           0                   0   
4      0                0         0           0                   0   

   Bike Rental / Bike Share  Bike Shop  Bistro  Board Shop  Boat or Ferry  \
0                         0          0       0           0              0   
1                         0          0       0           0              0   
2                         0          0       0           0              0   
3                         0          0       0           0              0   
4                         0          0       0           0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

In [32]:
# Examine the new data frame
toronto_onehot.shape

(2268, 275)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category.

In [33]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                              Adelaide,King,Richmond     0.000000   
1                                           Agincourt     0.000000   
2   Agincourt North,L'Amoreaux East,Milliken,Steel...     0.000000   
3   Albion Gardens,Beaumond Heights,Humbergate,Jam...     0.000000   
4                               Alderwood,Long Branch     0.000000   
5       Bathurst Manor,Downsview North,Wilson Heights     0.000000   
6                                     Bayview Village     0.000000   
7                    Bedford Park,Lawrence Manor East     0.000000   
8                                         Berczy Park     0.000000   
9                          Birch Cliff,Cliffside West     0.000000   
10  Bloordale Gardens,Eringate,Markland Wood,Old B...     0.000000   
11         Brockton,Exhibition Place,Parkdale Village     0.000000   
12  Business Reply Mail Processing Centre 969 Eastern     0.052632   
13                         CFB Toronto,Downsview East     0.000000   
14  CN Tower,Bathurst Quay,Island airport,Harbourf...     0.000000   
15                         Cabbagetown,St. James Town     0.000000   
16                                Caledonia-Fairbanks     0.000000   
17              Canada Post Gateway Processing Centre     0.000000   
18                                          Cedarbrae     0.000000   
19                                 Central Bay Street     0.011628   
20            Chinatown,Grange Park,Kensington Market     0.000000   
21                                           Christie     0.000000   
22                               Church and Wellesley     0.011111   
23                      Clairlea,Golden Mile,Oakridge     0.000000   
24              Clarks Corners,Sullivan,Tam O'Shanter     0.000000   
25      Cliffcrest,Cliffside,Scarborough Village West     0.000000   
26  Cloverdale,Islington,Martin Grove,Princess Gar...     0.000000   
27                      Commerce Court,Victoria Hotel     0.000000   
28                                         Davisville     0.000000   
29                                   Davisville North     0.000000   
30  Deer Park,Forest Hill SE,Rathnelly,South Hill,...     0.000000   
31        Del Ray,Keelesdale,Mount Dennis,Silverthorn     0.000000   
32            Design Exchange,Toronto Dominion Centre     0.000000   
33                                    Don Mills North     0.000000   
34  Dorset Park,Scarborough Town Centre,Wexford He...     0.000000   
35                        Dovercourt Village,Dufferin     0.000000   
36                                  Downsview Central     0.000000   
37                                Downsview Northwest     0.000000   
38                                     Downsview West     0.000000   
39                   Downsview,North Park,Upwood Park     0.000000   
40          East Birchmount Park,Ionview,Kennedy Park     0.000000   
41                                       East Toronto     0.000000   
42                                    Emery,Humberlea     0.000000   
43                         Fairview,Henry Farm,Oriole     0.000000   
44              First Canadian Place,Underground city     0.000000   
45                    Flemingdon Park,Don Mills South     0.000000   
46                 Forest Hill North,Forest Hill West     0.000000   
47                                          Glencairn     0.000000   
48                    Guildwood,Morningside,West Hill     0.000000   
49                      Harbord,University of Toronto     0.028571   
50    Harbourfront East,Toronto Islands,Union Station     0.000000   
51                           Harbourfront,Regent Park     0.019608   
52                       High Park,The Junction South     0.000000   
53               Highland Creek,Rouge Hill,Port Union     0.000000   
54                                  Hillcrest Village     0.000000   
55         Humber Bay Shores,Mimico South,New Toronto     0.000000   
56  Humber Bay,King's Mill Park,Kingsway Park Sout.

In [34]:
# confirm the new size
toronto_grouped.shape

(100, 275)

Let's print each neighborhood along with the top 5 most common venues.

In [35]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue', 'freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending = False).reset_index(drop = True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
            venue  freq
0     Coffee Shop  0.08
1            Café  0.05
2      Steakhouse  0.04
3             Bar  0.04
4  Cosmetics Shop  0.03


----Agincourt----
            venue  freq
0          Lounge  0.25
1  Breakfast Spot  0.25
2    Skating Rink  0.25
3  Clothing Store  0.25
4     Yoga Studio  0.00


----Agincourt North,L'Amoreaux East,Milliken,Steeles East----
                 venue  freq
0           Playground   0.5
1                 Park   0.5
2          Yoga Studio   0.0
3   Mexican Restaurant   0.0
4  Monument / Landmark   0.0


----Albion Gardens,Beaumond Heights,Humbergate,Jamestown,Mount Olive,Silverstone,South Steeles,Thistletown----
                  venue  freq
0           Pizza Place   0.2
1         Grocery Store   0.2
2              Pharmacy   0.1
3  Fast Food Restaurant   0.1
4        Sandwich Place   0.1


----Alderwood,Long Branch----
          venue  freq
0   Pizza Place  0.22
1  Skating Rink  0.11
2   Coffee Shop  0.11
3         

               venue  freq
0     Discount Store   0.4
1        Bus Station   0.2
2   Department Store   0.2
3        Coffee Shop   0.2
4  Mobile Phone Shop   0.0


----East Toronto----
                       venue  freq
0                Coffee Shop  0.33
1          Convenience Store  0.33
2                       Park  0.33
3                Yoga Studio  0.00
4  Middle Eastern Restaurant  0.00


----Emery,Humberlea----
                             venue  freq
0           Furniture / Home Store   0.5
1                   Baseball Field   0.5
2                      Yoga Studio   0.0
3        Middle Eastern Restaurant   0.0
4  Molecular Gastronomy Restaurant   0.0


----Fairview,Henry Farm,Oriole----
                  venue  freq
0        Clothing Store  0.15
1           Coffee Shop  0.08
2  Fast Food Restaurant  0.08
3         Women's Store  0.03
4                Bakery  0.03


----First Canadian Place,Underground city----
         venue  freq
0  Coffee Shop  0.10
1         Café  0.08
2    

                       venue  freq
0             Clothing Store  0.08
1                Coffee Shop  0.08
2                       Café  0.03
3             Cosmetics Shop  0.03
4  Middle Eastern Restaurant  0.03


----Scarborough Village----
                             venue  freq
0                       Playground   1.0
1                      Yoga Studio   0.0
2               Mexican Restaurant   0.0
3              Monument / Landmark   0.0
4  Molecular Gastronomy Restaurant   0.0


----St. James Town----
                venue  freq
0         Coffee Shop  0.07
1                Café  0.06
2          Restaurant  0.05
3               Hotel  0.05
4  Italian Restaurant  0.04


----Stn A PO Boxes 25 The Esplanade----
                venue  freq
0         Coffee Shop  0.11
1  Italian Restaurant  0.04
2          Restaurant  0.04
3                Café  0.04
4  Seafood Restaurant  0.03


----Studio District----
                 venue  freq
0                 Café  0.11
1          Coffee Shop  0.0

Let's put this into a pandas dataframe. First, let's write a function to sort the venues in descending order.

In [36]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending = False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [37]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind + 1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind + 1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns = columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                          Agincourt                Lounge   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...                  Park   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...           Pizza Place   
4                              Alderwood,Long Branch           Pizza Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café                   Bar            Steakhouse   
1          Skating Rink        Clothing Store        Breakfast Spot   
2            Playground         Women's Store             Drugstore   
3         Grocery Store              Pharmacy   Fried Chicken Joint   
4                   Gym                  Pool           Coffee Shop   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Restaurant                 Hotel        Cosmetics Shop   
1           Event Space    Falafel Restaurant        Farmers Market   
2    Dim Sum Restaurant                 Diner        Discount Store   
3           Coffee Shop        Sandwich Place            Beer Store   
4              Pharmacy          Skating Rink                   Pub   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0          Burger Joint       Thai Restaurant    American Restaurant  
1  Ethiopian Restaurant     Electronics Store     Dim Sum Restaurant  
2               Dog Run      Doner Restaurant             Donut Shop  
3  Fast Food Restaurant             Gift Shop            Curling Ice  
4        Sandwich Place          Dessert Shop     Dim Sum Restaurant

###  Clustering Neighborhoods

Run k-means to cluster the neighborhood into 5 clusters.

In [38]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters = kclusters, random_state = 0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 2, 0, 0, 0, 0, 0, 0, 0])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [41]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = neighborhoods

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on = 'Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode      Borough                          Neighborhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.806686   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  Cluster Labels       1st Most Common Venue  \
0 -79.194353             3.0        Fast Food Restaurant   
1 -79.160497             0.0  Construction & Landscaping   
2 -79.188711             0.0          Mexican Restaurant   
3 -79.216917             0.0                 Coffee Shop   
4 -79.239476             0.0            Hakka Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Women's Store   Dumpling Restaurant                 Diner   
1                   Bar         Women's Store   Dumpling Restaurant   
2   Rental Car Location     Electronics Store           Pizza Place   
3     Korean Restaurant   Dumpling Restaurant                 Diner   
4       Thai Restaurant    Athletics & Sports                Bakery   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Discount Store               Dog Run      Doner Restaurant   
1        Discount Store               Dog Run      Doner Restaurant   
2          Intersection        Breakfast Spot        Medical Center   
3        Discount Store               Dog Run      Doner Restaurant   
4                  Bank  Caribbean Restaurant   Fried Chicken Joint   

         8th Most Common Venue 9th Most Common Venue  \
0                   Donut Shop             Drugstore   
1                   Donut Shop             Drugstore   
2                Women's Store                 Diner   
3                   Donut Shop             Drugstore   
4  Eastern European Restaurant   Dumpling Restaurant   

        10th Most Common Venue  
0  Eastern European Restaurant  
1  Eastern European Restaurant  
2               Discount Store  
3                Women's Store  
4                    Drugstore

In [42]:
# create map
map_clusters = folium.Map(location = [latitude, longitude], zoom_start = 11)

toronto_merged['Cluster Labels'].fillna(5, inplace = True)  #replacing NaN with 5, which represents a new cluster

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i * x) ** 2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html = True)
    folium.CircleMarker(
        [lat, lon],
        radius = 5,
        popup = label,
        color = rainbow[int(cluster - 1)],
        fill = True,
        fill_color = rainbow[int(cluster - 1)],
        fill_opacity = 0.7).add_to(map_clusters)
       
map_clusters

### Examining Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster.

<b>Cluster 1

In [43]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
1         Scarborough             0.0  Construction & Landscaping   
2         Scarborough             0.0          Mexican Restaurant   
3         Scarborough             0.0                 Coffee Shop   
4         Scarborough             0.0            Hakka Restaurant   
6         Scarborough             0.0              Discount Store   
7         Scarborough             0.0                    Bus Line   
8         Scarborough             0.0                       Motel   
9         Scarborough             0.0             College Stadium   
10        Scarborough             0.0           Indian Restaurant   
11        Scarborough             0.0                  Smoke Shop   
12        Scarborough             0.0                      Lounge   
13        Scarborough             0.0                 Pizza Place   
15        Scarborough             0.0        Fast Food Restaurant   
17         North York             0.0                        Pool   
18         North York             0.0              Clothing Store   
19         North York             0.0                        Café   
21         North York             0.0                   Piano Bar   
22         North York             0.0                 Coffee Shop   
24         North York             0.0                    Pharmacy   
26         North York             0.0                        Café   
27         North York             0.0                 Coffee Shop   
28         North York             0.0                 Coffee Shop   
29         North York             0.0          Falafel Restaurant   
31         North York             0.0               Grocery Store   
32         North York             0.0              Baseball Field   
33         North York             0.0          Athletics & Sports   
34         North York             0.0       Portuguese Restaurant   
35          East York             0.0        Fast Food Restaurant   
36          East York             0.0                    Pharmacy   
37       East Toronto             0.0        Other Great Outdoors   
38          East York             0.0                 Coffee Shop   
39          East York             0.0           Indian Restaurant   
41       East Toronto             0.0            Greek Restaurant   
42       East Toronto             0.0              Sandwich Place   
43       East Toronto             0.0                        Café   
45    Central Toronto             0.0                       Hotel   
46    Central Toronto             0.0                 Coffee Shop   
47    Central Toronto             0.0                 Pizza Place   
48    Central Toronto             0.0                  Restaurant   
49    Central Toronto             0.0                 Coffee Shop   
51   Downtown Toronto             0.0                 Coffee Shop   
52   Downtown Toronto             0.0                 Coffee Shop   
53   Downtown Toronto             0.0                 Coffee Shop   
54   Downtown Toronto             0.0                 Coffee Shop   
55   Downtown Toronto             0.0                 Coffee Shop   
56   Downtown Toronto             0.0                 Coffee Shop   
57   Downtown Toronto             0.0                 Coffee Shop   
58   Downtown Toronto             0.0                 Coffee Shop   
59   Downtown Toronto             0.0                 Coffee Shop   
60   Downtown Toronto             0.0                 Coffee Shop   
61   Downtown Toronto             0.0                 Coffee Shop   
62         North York             0.0                 Coffee Shop   
63    Central Toronto             0.0     Health & Beauty Service   
65    Central Toronto             0.0                        Café   
66   Downtown Toronto             0.0                        Café   
67   Downtown Toronto             0.0                        Café   
68   Downtown Toronto             0.0            Airport Terminal   
69   Downtow

<b>Cluster 2

In [44]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
98    York             1.0                  Park         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
98   Dumpling Restaurant                 Diner        Discount Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
98               Dog Run      Doner Restaurant            Donut Shop   

   9th Most Common Venue       10th Most Common Venue  
98             Drugstore  Eastern European Restaurant

<b>Cluster 3

In [45]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
5   Scarborough             2.0            Playground         Women's Store   
14  Scarborough             2.0                  Park            Playground   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
5              Drugstore    Dim Sum Restaurant                 Diner   
14         Women's Store             Drugstore    Dim Sum Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
5         Discount Store               Dog Run      Doner Restaurant   
14                 Diner        Discount Store               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
5             Donut Shop    Dumpling Restaurant  
14      Doner Restaurant             Donut Shop

<b>Cluster 4

In [46]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  Scarborough             3.0  Fast Food Restaurant         Women's Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0   Dumpling Restaurant                 Diner        Discount Store   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0               Dog Run      Doner Restaurant            Donut Shop   

  9th Most Common Venue       10th Most Common Venue  
0             Drugstore  Eastern European Restaurant

<b>Cluster 5

In [47]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
23         North York             4.0                  Park   
25         North York             4.0     Food & Drink Shop   
30         North York             4.0               Airport   
40          East York             4.0     Convenience Store   
44    Central Toronto             4.0              Bus Line   
50   Downtown Toronto             4.0                  Park   
64    Central Toronto             4.0                 Trail   
72         North York             4.0                  Park   
74               York             4.0                  Park   
79         North York             4.0         Deli / Bodega   
90          Etobicoke             4.0                  Park   
100         Etobicoke             4.0              Bus Line   

    2nd Most Common Venue 3rd Most Common Venue       4th Most Common Venue  \
23                   Bank     Convenience Store               Women's Store   
25                   Park         Women's Store                       Diner   
30                   Park         Women's Store         Dumpling Restaurant   
40            Coffee Shop                  Park         Dumpling Restaurant   
44                   Park           Swim School               Women's Store   
50                  Trail            Playground                    Building   
64          Jewelry Store                  Park            Sushi Restaurant   
72            Pizza Place   Japanese Restaurant                         Pub   
74          Women's Store                Market        Fast Food Restaurant   
79                   Park                Bakery  Construction & Landscaping   
90             Smoke Shop                 River               Women's Store   
100           Pizza Place                  Park           Mobile Phone Shop   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
23    Dumpling Restaurant        Discount Store               Dog Run   
25         Discount Store               Dog Run      Doner Restaurant   
30                  Diner        Discount Store               Dog Run   
40                  Diner        Discount Store               Dog Run   
44             Donut Shop                 Diner        Discount Store   
50             Donut Shop    Dim Sum Restaurant                 Diner   
64                  Diner        Discount Store               Dog Run   
72          Women's Store      Doner Restaurant    Dim Sum Restaurant   
74     Falafel Restaurant           Event Space  Ethiopian Restaurant   
79          Women's Store             Drugstore        Discount Store   
90       Doner Restaurant          Dessert Shop    Dim Sum Restaurant   
100         Women's Store            Donut Shop                 Diner   

    8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
23       Doner Restaurant                   Donut Shop              Drugstore  
25             Donut Shop                    Drugstore    Dumpling Restaurant  
30       Doner Restaurant                   Donut Shop              Drugstore  
40       Doner Restaurant                   Donut Shop              Drugstore  
44                Dog Run             Doner Restaurant              Drugstore  
50         Discount Store                      Dog Run       Doner Restaurant  
64       Doner Restaurant                   Donut Shop              Drugstore  
72                  Diner               Discount Store                Dog Run  
74      Electronics Store  Eastern European Restaurant           Dessert Shop  
79                Dog Run             Doner Restaurant             Donut Shop  
90                  Diner               Discount Store                Dog Run  
100        Discount Store                      Dog Run       Doner Restaurant

This completes the assignment!